src/arithmetic/arithmetic_expression.c

```diff
@@ -361,86 +361,88 @@ SIValue AR_LEFT(SIValue *argv, int argc) {
+      left_str[newlen] = '\0';
     }
-
-    return SI_StringValC(left_str);
+    return SI_StringVal(left_str);
 }
 
 SIValue AR_LTRIM(SIValue *argv, int argc) {
     if(SIValue_IsNull(argv[0])) return SI_NullVal();
 
     assert(argc == 1 && argv[0].type == T_STRING);
     
-    char *str = argv[0].stringval.str;
-    int i = 0;
+    char *trimmed = argv[0].stringval;
 
-    while(str[i++] == ' ');
-    if(i>0) i--;
+    while(*trimmed == ' ') {
+      trimmed ++;
+    }
 
-    size_t trimmed_len = argv[0].stringval.len - i;
-    char *trimmed = malloc(sizeof(char) * trimmed_len);
+    char *ret = strdup(trimmed);
 
-    memcpy(trimmed, str+i, trimmed_len);
-    return SI_StringValC(trimmed);
+    return SI_StringVal(ret);
 }
 
 SIValue AR_RIGHT(SIValue *argv, int argc) {
     assert(argc == 2);
     if(SIValue_IsNull(argv[0])) return SI_NullVal();
     assert(argv[0].type == T_STRING);
     assert(argv[1].type == T_DOUBLE);
-    
-    char* right_str = strdup(argv[0].stringval.str);
-    size_t str_len = argv[0].stringval.len;
-    size_t right = (size_t)argv[1].doubleval;
-
-    /* Boundry check. */
-    if(right < str_len) {
-        /* TODO: free has to be called on original pointer position. */
-        right_str += str_len - right;
+
+    char *right_str;
+    int newlen = (int)argv[1].doubleval;
+    int start = strlen(argv[0].stringval) - newlen;
+
+    if (start <= 0) {
+      // No need to truncate this string based on the requested length
+      right_str = strdup(argv[0].stringval);
+    } else {
+      right_str = strdup(argv[0].stringval + start);
     }
 
-    return SI_StringValC(right_str);
+    return SI_StringVal(right_str);
 }
 
 SIValue AR_RTRIM(SIValue *argv, int argc) {
     if(SIValue_IsNull(argv[0])) return SI_NullVal();
     assert(argc == 1 && argv[0].type == T_STRING);
     
-    char *str = argv[0].stringval.str;
-    int i = argv[0].stringval.len-1;
+    char *str = argv[0].stringval;
+    int i = strlen(str) - 1;
     
     while(i >= 0 && str[i] == ' ') {
-         i--;
+      i --;
     }
     
     size_t trimmed_len = i+1; /* NULL terminator.*/
-    char *trimmed = malloc(sizeof(char) * (trimmed_len));
+    char *trimmed = malloc(sizeof(char) * trimmed_len);
 
     memcpy(trimmed, str, trimmed_len);
-    return SI_StringValC(trimmed);
+    trimmed[trimmed_len] = '\0';
+
+    return SI_StringVal(trimmed);
 }
 
 SIValue AR_REVERSE(SIValue *argv, int argc) {
     if(SIValue_IsNull(argv[0])) return SI_NullVal();
     assert(argv[0].type == T_STRING);
-    char *str = argv[0].stringval.str;
-    size_t str_len = argv[0].stringval.len;
+    char *str = argv[0].stringval;
+    size_t str_len = strlen(argv[0].stringval);
     char* reverse = malloc(sizeof(char) * (str_len+1));
     
     int i = str_len-1;
     int j = 0;
     while(i >= 0) { reverse[j++] = str[i--]; }
     reverse[j] = 0;
-    return SI_StringValC(reverse);
+    return SI_StringVal(reverse);
```

Comment:
Need to free reverse as SI_StringVal duplicates.
*Maybe allocate reverse on the stack? (if so this applies to all other review comments above)